rcl_yaml_param_parser/src/parse.c

```diff
@@ -404,6 +409,147 @@ rcutils_ret_t parse_value(
+    return RCUTILS_RET_ERROR;
+  }
+  if (RMW_NODE_NAME_VALID != validation_result) {
+    RCUTILS_SET_ERROR_MSG(rmw_node_name_validation_result_string(validation_result));
+    return RCUTILS_RET_ERROR;
+  }
+
+  return RCUTILS_RET_OK;
+}
+
+///
+/// Check a name (namespace/node_name) whether it is valid
+///
+static rcutils_ret_t
+__validate_name(const char * name, rcutils_allocator_t allocator)
+{
+  // special rules
+  if (0 == strcmp(name, "/**") || 0 == strcmp(name, "/*")) {
+    return RCUTILS_RET_OK;
+  }
+
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+  char * separator_pos = rindex(name, '/');
+  char * node_name = NULL;
+  char * absolute_namespace = NULL;
+  if (NULL == separator_pos) {
+    node_name = rcutils_strdup(name, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+  } else {
+    // substring namespace including the last '/'
+    char * namespace = rcutils_strndup(name, separator_pos - name + 1, allocator);
+    if (NULL == namespace) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+    if (namespace[0] != '/') {
+      absolute_namespace = rcutils_format_string(allocator, "/%s", namespace);
+      allocator.deallocate(namespace, allocator.state);
+      if (NULL == absolute_namespace) {
+        ret = RCUTILS_RET_BAD_ALLOC;
+        goto clean;
+      }
+    } else {
+      absolute_namespace = namespace;
+    }
+
+    node_name = rcutils_strdup(separator_pos + 1, allocator);
+    if (NULL == node_name) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+      goto clean;
+    }
+  }
+
+  if (absolute_namespace) {
+    size_t i = 0;
+    separator_pos = index(absolute_namespace + i + 1, '/');
+    if (NULL == separator_pos) {
+      ret = __validate_namespace(absolute_namespace);
+      if (RCUTILS_RET_OK != ret) {
+        goto clean;
+      }
+    } else {
+      do {
+        size_t len = separator_pos - absolute_namespace - i;
+        char * namespace = rcutils_strndup(absolute_namespace + i, len, allocator);
+        if (NULL == namespace) {
+          ret = RCUTILS_RET_BAD_ALLOC;
+          goto clean;
+        }
+        if (0 == strcmp(namespace, "/")) {
+          RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+            "%s contains repeated forward slash", absolute_namespace);
+          allocator.deallocate(namespace, allocator.state);
+          ret = RCUTILS_RET_INVALID_ARGUMENT;
+          goto clean;
+        }
+        if (0 != strcmp(namespace, "/**") && 0 != strcmp(namespace, "/*")) {
+          ret = __validate_namespace(namespace);
+          if (RCUTILS_RET_OK != ret) {
+            allocator.deallocate(namespace, allocator.state);
+            goto clean;
+          }
+        }
+        allocator.deallocate(namespace, allocator.state);
+        i += len;
+      } while (NULL != (separator_pos = index(absolute_namespace + i + 1, '/')));
+    }
+  }
+
+  if (0 != strcmp(node_name, "*") && 0 != strcmp(node_name, "**")) {
```

Comment:
I don't think a node name set to `**` should be valid, since it implies it has zero or more parts (eg. `""` or `"/foo/bar"`).

Reply:
This case was for '/foo/bar/**'. OK, I'll add more codes to not allow node name to be '**'.

Reply:
I see. Yeah, this is tricky case. Maybe it's fine to leave the code as-is.

Reply:
okay.